File: test/extensions/dynamic_modules/test_data/rust/upstream_http_tcp_bridge.rs
use envoy_proxy_dynamic_modules_rust_sdk::*;

declare_upstream_http_tcp_bridge_init_functions!(program_init, new_bridge_config);

fn program_init() -> bool {
  true
}

fn new_bridge_config(_name: &str, config: &[u8]) -> Option<Box<dyn UpstreamHttpTcpBridgeConfig>> {
  let config_str = std::str::from_utf8(config).unwrap_or("streaming");
  let mode = match config_str {
    "local_reply" => BridgeMode::LocalReply,
    _ => BridgeMode::Streaming,
  };
  Some(Box::new(TestBridgeConfig { mode }))
}

#[derive(Clone, Copy)]
enum BridgeMode {
  Streaming,
  LocalReply,
}

struct TestBridgeConfig {
  mode: BridgeMode,
}

impl UpstreamHttpTcpBridgeConfig for TestBridgeConfig {
  fn new_bridge(
    &self,
    _envoy_bridge: &dyn EnvoyUpstreamHttpTcpBridge,
  ) -> Box<dyn UpstreamHttpTcpBridge> {
    Box::new(TestBridge {
      mode: self.mode,
      response_headers_sent: false,
    })
  }
}

struct TestBridge {
  mode: BridgeMode,
  response_headers_sent: bool,
}

impl UpstreamHttpTcpBridge for TestBridge {
  fn on_encode_headers(
    &mut self,
    envoy_bridge: &dyn EnvoyUpstreamHttpTcpBridge,
    _end_of_stream: bool,
  ) {
    match self.mode {
      BridgeMode::Streaming => {
        if let (Some(method), _) = envoy_bridge.get_request_header_value(":method", 0) {
          let method_str = std::str::from_utf8(&method).unwrap_or("?");
          let prefix = format!("METHOD={method_str} ");
          envoy_bridge.send_upstream_data(prefix.as_bytes(), false);
        }
      },
      BridgeMode::LocalReply => {
        envoy_bridge.send_response(403, &[], b"access denied");
      },
    }
  }

  fn on_encode_data(&mut self, envoy_bridge: &dyn EnvoyUpstreamHttpTcpBridge, end_of_stream: bool) {
    let data = envoy_bridge.get_request_buffer();
    envoy_bridge.send_upstream_data(&data, end_of_stream);
  }

  fn on_encode_trailers(&mut self, envoy_bridge: &dyn EnvoyUpstreamHttpTcpBridge) {
    envoy_bridge.send_upstream_data(&[], true);
  }

  fn on_upstream_data(
    &mut self,
    envoy_bridge: &dyn EnvoyUpstreamHttpTcpBridge,
    end_of_stream: bool,
  ) {
    if !self.response_headers_sent {
      envoy_bridge.send_response_headers(200, &[("x-bridge-mode", b"dynamic_module")], false);
      self.response_headers_sent = true;
    }

    let tcp_data = envoy_bridge.get_response_buffer();
    envoy_bridge.send_response_data(&tcp_data, end_of_stream);
  }
}

File: test/extensions/dynamic_modules/test_data/go/upstream_http_tcp_bridge/upstream_http_tcp_bridge.go
// Upstream HTTP/TCP bridge test module. Mirrors test_data/rust/upstream_http_tcp_bridge.rs.
//
// Two modes selected by the config bytes:
//
//	"local_reply" — short-circuits with a 403 local reply on encode_headers.
//	anything else — streaming bridge: forwards request method as a "METHOD=X " prefix to
//	                the upstream, then forwards the request body, and converts upstream
//	                bytes back into HTTP response data.
//
// Loaded by test/extensions/upstreams/http/dynamic_modules/integration_test.cc which is
// parameterized over (rust, go).
package main

import (
	"fmt"

	sdk "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go"
	_ "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/abi"
	"github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/shared"
)

func init() {
	sdk.RegisterUpstreamHttpTcpBridgeConfigFactories(map[string]shared.UpstreamHttpTcpBridgeConfigFactory{
		"test_bridge": &testBridgeConfigFactory{},
	})
}

func main() {} //nolint:all

type bridgeMode int

const (
	modeStreaming bridgeMode = iota
	modeLocalReply
)

type testBridgeConfigFactory struct{}

func (testBridgeConfigFactory) Create(_ string, config []byte) (shared.UpstreamHttpTcpBridgeFactory, error) {
	mode := modeStreaming
	if string(config) == "local_reply" {
		mode = modeLocalReply
	}
	return &testBridgeFactory{mode: mode}, nil
}

type testBridgeFactory struct {
	shared.EmptyUpstreamHttpTcpBridgeFactory
	mode bridgeMode
}

func (f *testBridgeFactory) Create(_ shared.UpstreamHttpTcpBridgeHandle) shared.UpstreamHttpTcpBridge {
	return &testBridge{mode: f.mode}
}

type testBridge struct {
	shared.EmptyUpstreamHttpTcpBridge
	mode                bridgeMode
	responseHeadersSent bool
}

func (b *testBridge) EncodeHeaders(handle shared.UpstreamHttpTcpBridgeHandle, _ bool) {
	switch b.mode {
	case modeStreaming:
		method, _, ok := handle.GetRequestHeader(":method", 0)
		if ok {
			prefix := fmt.Sprintf("METHOD=%s ", method.ToUnsafeString())
			handle.SendUpstreamData([]byte(prefix), false)
		}
	case modeLocalReply:
		handle.SendResponse(403, nil, []byte("access denied"))
	}
}

func (b *testBridge) EncodeData(handle shared.UpstreamHttpTcpBridgeHandle, endOfStream bool) {
	chunks := handle.GetRequestBuffer()
	if len(chunks) == 0 {
		if endOfStream {
			handle.SendUpstreamData(nil, true)
		}
		return
	}
	for i, chunk := range chunks {
		isLast := endOfStream && i == len(chunks)-1
		handle.SendUpstreamData(chunk.ToUnsafeBytes(), isLast)
	}
}

func (b *testBridge) EncodeTrailers(handle shared.UpstreamHttpTcpBridgeHandle) {
	handle.SendUpstreamData(nil, true)
}

func (b *testBridge) OnUpstreamData(handle shared.UpstreamHttpTcpBridgeHandle, endOfStream bool) {
	if !b.responseHeadersSent {
		handle.SendResponseHeaders(200,
			[][2]string{{"x-bridge-mode", "dynamic_module"}}, false)
		b.responseHeadersSent = true
	}
	chunks := handle.GetResponseBuffer()
	if len(chunks) == 0 {
		if endOfStream {
			handle.SendResponseData(nil, true)
		}
		return
	}
	for i, chunk := range chunks {
		isLast := endOfStream && i == len(chunks)-1
		handle.SendResponseData(chunk.ToUnsafeBytes(), isLast)
	}
}
